streamer: Use `BytesPacket` instead of `Packet`

vadorovsky opened the pull request:
#### Problem

The `Packet` struct stores the payload in a byte array. Using that struct often involves moving and cloning it, which then results in large memory copies.

#### Summary of Changes

`BytesPacket` is the new type of packet (introduced in anza-xyz/agave#5646) which uses `bytes::Bytes` and therefore doesn't make an actual copy of the underlying data. We use it in QUIC servers already.

Use it in the UDP streamer as well, to let the other components of Agave (gossip, turbine etc.) benefit from the zero-copy capabilities.

<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

alexpyattaev commented on core/src/repair/repair_service.rs:
```diff
@@ -1310,16 +1312,14 @@ mod test {
-        let Some(bytes) = packet.data(..).map(Vec::from) else {
-            panic!("packet data not found");
-        };
+        let mut buffer = BytesMut::with_capacity(PACKET_DATA_SIZE);
```
This might allow >1 packet to get received if they are sufficiently small, while the original code only allowed for 1 packet to be received. Would it not be better to just call straight up recv on the socket and avoid all the silly dance with recv_mmsg? Would make code far cleaner.

vadorovsky replied:
> This might allow >1 packet to get received if they are sufficiently small, while the original code only allowed for 1 packet to be received.

It's true that the recvmmsg call in this unit test previously allowed max 1 packet to be received. In practice, these unit tests will receive just as many packets as there are being sent.

In theory, we could make `RecvMetas` configurable and just let the tests limit the maximum amount of packets that can be received. But I don't see a point. Such mechanism has no use in production (where we always want to receive bigger batches) and tests are working just fine, even though they *could* receive 64 packets. It would add churn to the code without visible benefits IMO.

To be precise, the only real user of our `recv_mmsg` (which is then used by gossip, turbine etc.) is here:

https://github.com/vadorovsky/agave/blob/packet-bytes-streamer/streamer/src/streamer.rs#L152-L232
https://github.com/vadorovsky/agave/blob/packet-bytes-streamer/streamer/src/packet.rs#L30-L104

It's made for 64 packet batches only. 

>Would it not be better to just call straight up recv on the socket and avoid all the silly dance with recv_mmsg? Would make code far cleaner.

But then such a test would test Rust stdlib instead of our `recv_mmsg` utility.

If you are suggesting using recv on the socket **in the real server code** (gossip, turbine), not just in tests, then I feel even more strongly against that, because we could receive only one packet per call. So we would end up:

* Issuing a syscall 64 times, instead of doing that once, like we do now
* Allocating small buffers 64 times, instead of doing that once per batch.

My goal here is to reduce allocations and syscalls.

In fact, one of the improvements I'm trying to add here (in comparison with current master) is making a buffer allocation once here:

https://github.com/vadorovsky/agave/blob/packet-bytes-streamer/streamer/src/streamer.rs#L162
https://github.com/vadorovsky/agave/blob/packet-bytes-streamer/streamer/src/recvmmsg.rs#L28-L31

(I need to get rid of the `zeroed` allocation and make the rest of the code work with an uninitialized buffer, I want to fix that before marking as ready)

Outside of the outer recv loop. That big buffer is allocated once for the whole 64 packet batch, and then being split into smaller:

* `&mut [u8]` instances for receiving
* `Bytes` instances for using as `BytesPacket` in the caller

Then that buffer is potentially resized here:

https://github.com/vadorovsky/agave/blob/packet-bytes-streamer/streamer/src/streamer.rs#L226

But if we are lucky and the `BytesPacket` instances backed by that buffer are already consumed by the caller before the next loop iteration, this `reserve()` call might not do any real resize/reallocation. In the worst case, it will fully reallocate the buffer, but that's unavoidable in the pessimistic case.

On master, we allocate multiple ~1kb buffers in the inner loop:

https://github.com/anza-xyz/agave/blob/b3a310c564fe81c2c04d330ddd55607fa41c6c30/streamer/src/streamer.rs#L190-L195

Which is still worse (it's always doing more allocations).

alexpyattaev commented on gossip/src/cluster_info.rs:
```diff
@@ -217,11 +217,10 @@ impl ClusterInfo {
         gossip_validators: Option<&HashSet<Pubkey>>,
         sender: &impl ChannelSend<PacketBatch>,
-    ) {
+    ) -> Result<(), GossipError> {
```
What is the purposed of adding error handling here if we are going to unwrap the error at the callsite? This is potentially taking a code that will silently ignore errors (which is not great but not terrible) and turning it into a code which can easily panic at runtime.

gregcusack replied:
ahh looks like `make_gossip_packet_batch` has to now return an error which gets propagated up to `send_gossip_packets()`. but you make a good point, we shouldn't `.unwrap()` in the caller.

vadorovsky replied:
Good point, I will remove the error handling changes. More details in the comment below. :sweat_smile:

alexpyattaev commented on gossip/src/cluster_info.rs:
```diff
@@ -2032,26 +2020,14 @@ impl ClusterInfo {
             .map(|(addr, ping)| (addr, Protocol::PingMessage(ping)));
-        send_gossip_packets(pings, recycler, response_sender, &self.stats);
-        self.handle_batch_ping_messages(ping_messages, recycler, response_sender);
+        send_gossip_packets(pings, response_sender, &self.stats)?;
```
Same here, let us not change error handling patterns in gossip in this PR to keep the scope narrow.

vadorovsky replied:
The current code on main ends up using `Packet::populate_packet`, which also returns a `Result`, but it is logged  in this function:

https://github.com/anza-xyz/agave/blob/ddefc131edf65b3ca5dfcdfc2af02e420834b5db/perf/src/packet.rs#L684

I guess I could write a similar method for `BytesPacketBatch` and stop sprinkling the error in gossip.

I didn't do that before, because I felt like ignoring a serialization error is a smell, but I agree with you that it adds unnecessary diff and that we should propose error handling changes as separate PRs. So I will just remove the error handling changes in the next push.

alexpyattaev commented on ledger/src/shred.rs:
```diff
@@ -1040,6 +1046,37 @@ pub fn verify_test_data_shred(
     }
 }
 
+#[cfg(feature = "dev-context-only-utils")]
```
Can we use an already public `fn make_merkle_shreds_from_entries` with similar semantics instead of exposing this function?

vadorovsky replied:
I can try. BTW, this is a change related to this prerequisite PR: #6735. I will fix that there as well.

alexpyattaev commented on streamer/src/nonblocking/recvmmsg.rs:
```diff
@@ -16,7 +13,7 @@ pub async fn recv_mmsg(
 ) -> io::Result</*num packets:*/ usize> {
     debug_assert!(packets.iter().all(|pkt| pkt.meta() == &Meta::default()));
-    let count = cmp::min(NUM_RCVMMSGS, packets.len());
+    let count = packets.len();
```
why is it safe to get rid of the `min` here?

vadorovsky replied:
The callers of this function are never making batches bigger than this constant. Having this check is redundant.

If you really insist, I could turn it into `debug_assert!`, to catch some potential mistake of calling this function with too big batches. But doing that comparison in production on every packet receive. There is a slight chance that LLVM passes optimize it out anyway. :joy: 

Also, I removed the `NUM_RCVMMSG` constant, because it just repeats `PACKETS_PER_BATCH`.

I can propose that as a separate PR, to remove the diff here.

alexpyattaev replied:
Ok quick check implies this function is never actually used in agave outside of tests. I'll go plant a DCOU on it.

vadorovsky replied:
The same code is in other places though. I'm gonna clean it up in separate PRs.

I've already submitted one for removing the reduntant `NUM_RCMMMSGS` constant: #6955

alexpyattaev reviewed (commented):
This looks very promising, thank you! My only serious concern with this is in gossip where you end up modifying error handling logic. I would advise against doing that in a big PR like this. Gossip is rather tricky to reason about, so if it is required to change error handling there, I'd suggest it be done in a separate PR.

gregcusack commented on gossip/src/cluster_info.rs:
```diff
@@ -1480,12 +1476,8 @@ impl ClusterInfo {
-                            &sender,
-                        );
+                        self.refresh_push_active_set(&stakes, gossip_validators.as_ref(), &sender)
+                            .unwrap();
```
can we handle this error better? probably don't want to panic if we fail to send out some pings.

vadorovsky replied:
I will just remove this diff. This is yet again because of `send_gossip_packets`/`BytesPacket::from_data` returning an error. I will just log the error when sending packets.

gregcusack commented on gossip/src/cluster_info.rs:
```diff
@@ -237,7 +236,8 @@ impl ClusterInfo {
             .into_iter()
             .map(|(addr, ping)| (addr, Protocol::PingMessage(ping)));
-        send_gossip_packets(pings, recycler, sender, &self.stats);
+        send_gossip_packets(pings, sender, &self.stats)?;
```
ya i would actually handle the error here rather than up in `gossip()` and then don't return anything. `gossip()` doesn't need to know about a send error.

gregcusack reviewed (commented):
fly by review on the gossip stuff. looks good from a gossip logic perspective, just needs some updated error handling

vadorovsky commented on gossip/src/cluster_info.rs:
```diff
@@ -3244,14 +3210,14 @@ mod tests {
+                    remote_nodes.iter().map(|(_, socket)| socket).zip(pings),
+                    &sender,
+                )
+                .unwrap();
```
The send is going to fail only if the `Protocol` fails to serialize. Very unlikely that it will happen. But, again, for the sake of making the dill small, I will log the error somewhere else and stop sprinkling it over gossip. :slightly_smiling_face:

vadorovsky committed:
streamer: Use `BytesPacket` instead of `Packet`

The `Packet` struct stores the payload in a byte array. Using that
struct often involves moving and cloning it, which then results in
large memory copies.

BytesPacket` is the new type of packet (introduced in #5646) which uses
`bytes::Bytes` and therefore doesn't make an actual copy of the
underlying data. We use it in QUIC servers already.

Use it in the UDP streamer as well, to let the other components of Agave
(gossip, turbine etc.) benefit from the zero-copy capabilities.

alexpyattaev commented on streamer/src/recvmmsg.rs:
```diff
@@ -13,30 +13,40 @@ use {
 #[cfg(not(target_os = "linux"))]
-pub fn recv_mmsg(socket: &UdpSocket, packets: &mut [Packet]) -> io::Result</*num packets:*/ usize> {
-    debug_assert!(packets.iter().all(|pkt| pkt.meta() == &Meta::default()));
+pub fn recv_mmsg(
```
Maybe worth documenting that this will in fact append to provided BytesPacketBatch up to its capacity? This is not quite how the old function behaved, would not want someone to accidentally stumble over this one.

alexpyattaev reviewed (commented):
LGTM for the most part, left a few nits and suggestions.

alexpyattaev reviewed (commented):
Made another pass around unsafe blocks, there is technically UB in one place (though it probably is inconsequential in practice). We are technically casting uninitialized memory into a mutable slice which is "not allowed".

KirillLykov reviewed (commented):
Looks good to me overal

vadorovsky commented on streamer/src/recvmmsg.rs:
```diff
@@ -13,30 +13,40 @@ use {
+    let remaining_capacity = packets.capacity().saturating_sub(packets.len());
+    let count = cmp::min(PACKETS_PER_BATCH, remaining_capacity);
+    for _ in 0..count {
+        let mut buffer = BytesMut::with_capacity(PACKET_DATA_SIZE);
```
Good call, I can just allocate zeroed memory here, since we don't care about performance outside Linux.

If I wanted to be pedantic about perf, I could issue the raw syscall and pass the pointer, but seems like an overkill.

alexpyattaev reviewed (changes requested):
Need to sort out the set_nonblocking flag in packet.rs
